### ISSUE TITLE
Shadow DOM and Mobile View issue

### ISSUE THREAD

**Issue by skuzmin** [jackocnr/intl-tel-input#1481]
### Steps to reproduce
1. Create basic [WebComponent](https://developer.mozilla.org/en-US/docs/Web/API/Web_components) with intl-tel-input plugin
2. Switch to mobile view and refresh page (enable mobile mode)
3. Click open list of countries

### Expected behaviour
Pop-up(modal/dialog) with list of countries should be visible

### Actual behaviour
There is no Pop-up(modal/dialog). Because in mobile mode we have "iti-mobile" class on document.body(hardcoded)
<img width="353" alt="image" src="https://github.com/jackocnr/intl-tel-input/assets/8438624/48d30476-ce59-46c1-9f5c-01d210714f44">
And this doesn't have effect on incapsulated content inside shadowDom.

The same story with container, but in this case I can specify container inside shadowDOM. It would be great to have option to select element where we can add class for mobile view/mode in the same way as we have for container. If I add class manually to shadow dow element then everything works.

### Initialisation options
utilsScript: https://cdn.jsdelivr.net/npm/intl-tel-input@latest/build/js/utils.js
dropdownContainer: element inside shadow DOM

**Comment by jackocnr** [comment]
I see we reference `document.body` 5 times in the [source JS](https://github.com/jackocnr/intl-tel-input/blob/master/src/js/intlTelInput.js) so we would need to replace all of these. I propose a new bodyContainer option, which defaults to `document.body`. Would that solve your issue?

**Comment by jackocnr** [comment]
Well we're talking about the same thing - adding a new option to pass in a container element. You want this container element to take the iti-mobile class, but while we're doing that, we may as well use it for the other instances where we use document.body in the code, to keep it encapsulated within the shadow DOM. e.g. when you open the dropdown, the code currently uses document.body.scrollLeft to figure out where to position the dropdown, and in your case, this will need to be relative to the shadow DOM instead of the whole document.

I don't have time to implement this myself right now, but I would be open to a pull request.

**Comment by oprichnik** [comment]
Behaviour can be reproduced by changing user-agent in your browser.

If you don't need the popup, you can disable it with the `useFullscreenPopup` option.

**Comment by mimoudix** [comment]
did you fix this issue ? same btw

**Comment by skuzmin** [comment]
@mimoudix sorry ,I haven't had time to take a look into it, because of holidays :)

but I added temporary fix to my app (web component) to be able to use the lib in mobile view.
Maybe it will be helpful for you: 
add this into your code
`
      let dropdownContainer = undefined;
      const isMobile = /Android.+Mobile|webOS|iPhone|iPod|BlackBerry|IEMobile|Opera Mini/i.test(navigator.userAgent);
      if (isMobile) {
        // add class 'iti-mobile' to the upper element inside shadow dom (not the root)
        dropdownContainer = INPUT_WRAPPER_ELEMENT
      };

      intlTelInput(input, {
        utilsScript,
        dropdownContainer,
        separateDialCode: true,
      });
`

**Comment by mimoudix** [comment]
@skuzmin is it the same issue ? https://github.com/jackocnr/intl-tel-input/issues/1493

**Comment by skuzmin** [comment]
if you are using input-tel inside shadow dom , then yes... that's the same issue

if you are NOT using input-tel inside shadow dom, then it should be something different.

you can easily test mobile view on demo page, it works like a charm :) ( there is no shadow dom ofc)

**Comment by mimoudix** [comment]
@skuzmin  can you check from your side ? here is a live demo [profily.link/register](https://profily.link/register). Happy holidays !!

**Comment by skuzmin** [comment]
@mimoudix library was update and my issue might be fixed, but I will double check it after New Year.

About your issue, the problem is in styles (CSS).

So you need to update to the latest version and check all your styles ( to be the laster version as well)

In worst case scenario just add this styles to your global CSS:

<img width="394" alt="image" src="https://github.com/jackocnr/intl-tel-input/assets/8438624/6e71a489-6fc7-402a-9720-9ae675e1bcbe">
<img width="327" alt="image" src="https://github.com/jackocnr/intl-tel-input/assets/8438624/bb0211d0-4d39-427c-8319-66964fc0b3cd">

test on your web app:
<img width="1172" alt="image" src="https://github.com/jackocnr/intl-tel-input/assets/8438624/fa6cbbf1-93aa-4a71-bd4a-334ab1478bbe">


P.S. Don't forget to reload page when you switch from desktop to mobile view and back. The library doesn't detect mobile-desktop view change on window resize.

**Comment by mimoudix** [comment]
@skuzmin it worked, thanks!! Happy holidays.

**Comment by skuzmin** [comment]
@jackocnr I have create PR, please check when you have a time ( minor changes) 
https://github.com/jackocnr/intl-tel-input/pull/1497

**Comment by jackocnr** [comment]
Thanks for putting together the PR, but before I can merge it I need to understand exactly what the problem is here because re-reading our conversation I wonder if I've misunderstood.

My questions for you:

(1) Reviewing the plugin JS, there are several references to global variables. There are 35 references to `window`, 16 references to `document` and 1 reference to `navigator`. Will these references even work from inside a shadow DOM? (if not, the plugin will be hopelessly broken)

(2) Even if they do, my next thought is: does it really make sense to "encapsulate" this plugin in a web component / shadow DOM, when realistically it is not encapsulated at all! It is constantly leaking into the global scope. Do you not think this is a problem?

(3) If you don't think that is a problem, then I don't understand the need for this container class - why not just let the class be added to the document.body, and let the dropdown be appended to document.body (which it is by default, if useFullscreenPopup is true) and it should work fine, no?

(4) Which initialisation options are you using btw?

**Comment by skuzmin** [comment]
hi @jackocnr and sorry for delayed response,

1) window is available in the shadow dom, you can grab data from that object, but you can't set data inside the shadow dom.

2) I don't have a goal to encapsulate or something, I was forced to use MFE ( micro front ends ) on my project right now, and your library is the best in slot for working with phones, that's all.

3) So I will describe it in details. When I add MFE to a some web-site (let's call it CONTAINER), CONTAINER can't affect my MFE using JS or CSS.
**In case of desktop view**, I have all necessary css rules inside my MFE(shadow dom), and your lib works like a charm.
**In case of mobile view**, the lib moves the country list to the document.body and adds class to document.body, but STYLES for that are inside MFE(shadow root). So we have a case when the lib needs to handle something which exists out of its scope.
If the lib adds the country list and the class to shadowRoot.body instead of document.body it would work as well .
Maybe this abstract image will be helpful :
<img width="308" alt="image" src="https://github.com/jackocnr/intl-tel-input/assets/8438624/dfb41fd6-0738-40eb-9482-8c0b392aea6e">

long story short, shadowDom has access to window object ( for instance localstorage) but you can't affect shadow dom from the outside, so that's a reason why I can get your global objects, but styles from CONTAINER can't affect the MFE.
We have a property for the country list, so I decided to create similar for mobile class.

4) 
<img width="340" alt="image" src="https://github.com/jackocnr/intl-tel-input/assets/8438624/5b51b680-1ce8-4d1d-9b2c-d268adcad151">

P.S.
Hack which I'm using right now to resolve the issue (yeah Im using little be older version with "iti-mobile"):
<img width="954" alt="image" src="https://github.com/jackocnr/intl-tel-input/assets/8438624/bf2b698e-8cfd-436b-bb10-9f1ba17edfb8">

**Comment by jackocnr** [comment]
So if I understand correctly, the whole problem is that we put that one class on the body. I've just tried moving it instead to the dropdown container element, which makes more sense anyway. This change was released in v18.4.0 if you want to try it. Does that fix your problem?

**Comment by skuzmin** [comment]
In theory it should fix the issue, without any props and so one, you will need to rewrite CSS as well. Because if you add the class to the dropdown container this rule will not work:
<img width="234" alt="image" src="https://github.com/jackocnr/intl-tel-input/assets/8438624/1e136ce1-6226-409d-9480-b5bfbccf95d4">

I have just tested : 
<img width="357" alt="image" src="https://github.com/jackocnr/intl-tel-input/assets/8438624/e30f2509-527f-4f6e-aca1-6c8ab7e7f544">

**Comment by jackocnr** [comment]
Yeah I did all that 👍 

Give it a try and let me know if it works for you.

**Comment by skuzmin** [comment]
@jackocnr I can confirm that now it's working. Thx.

What should I do with PR? just close or you will close it ?

**Comment by skuzmin** [comment]
I have closed PR.


Issue has ben resolved.
